pkg/kwok/controllers/metrics.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"sigs.k8s.io/kwok/pkg/apis/internalversion"
+	kwokmetrics "sigs.k8s.io/kwok/pkg/kwok/metrics"
+	"sigs.k8s.io/kwok/pkg/kwok/metrics/cel"
+	"sigs.k8s.io/kwok/pkg/utils/maps"
+	"sigs.k8s.io/kwok/pkg/utils/slices"
+)
+
+var metricsHandler nodeMetricHandler
+
+func setMetrics(metrics []*internalversion.Metric) {
+	metricsHandler.metrics = metrics
+}
+
+type nodeMetricHandler struct {
+	// state
+	nodeContainersCounter maps.SyncMap[string, float64]
+	nodeCache             maps.SyncMap[string, *corev1.Node]
+
+	// configuration
+	metrics []*internalversion.Metric
+}
+
+func (h *nodeMetricHandler) updateCache(node *corev1.Node) {
+	h.nodeCache.Store(node.Name, node)
```

Comment:
Since metric updates can occur from the `NodeController` and the `PodController`, I thought it would be best to cache the latest node state and apply all metrics on publish. I wanted to avoid using a client to fetch the node in case `PodController` updates the number of containers running on a `Node`.
Does it make sense?